### PR TITLE
added support for escape sequences in r_num_as_string

### PR DESCRIPTION
**Pull request by bluec0re**
consists of 2 changes:

1. allow space in the original implementation as character
2. support escape sequences for nonprintable chars (uses currently hardcoded compile flag)

see #5896

### CODE REVIEW

**Comment by radare** [comment]
Here's the problem 
screen shot 2016-10-03 at 02 39 46

i would suggest to just do have two mode for this function, maybe adding another parameter for this, to discard or not printable strings to fix those tests and keep having the other functionality

File: libr/util/num.c [hunk 1]
```diff
@@ -408,17 +409,54 @@ R_API ut64 r_num_get_input_value(RNum *num, const char *input_value) {
...
 	ut64 num = n;
 	str[stri=0] = 0;
 	while (len--) {
 		char ch = (num & 0xff);
-		if (ch>=33 && ch <127) {
+		if (ch>=32 && ch <127) {
 			str[stri++] = ch;
 			str[stri] = 0;
+#ifdef R_NUM_AS_STRING_ESCAPES
+		} else if (ch == 7) {
```
Comment (radare) [inline libr/util/num.c]: looks spaguetti what about a switch table and a helper function to get only 1 char as argument and the string and returns the stri increment

File: libr/util/num.c [hunk 2]
```diff
@@ -408,17 +409,54 @@ R_API ut64 r_num_get_input_value(RNum *num, const char *input_value) {
 	return value;
 }
 
+#define NIBBLE_TO_HEX(n) (((n) & 0xf) > 9 ? 'a' + ((n) & 0xf) - 10 : '0' + ((n) & 0xf))
 R_API char* r_num_as_string(RNum *___, ut64 n) {
```
Comment (radare) [inline libr/util/num.c]: add parameter to toggle this functionality

Comment (bluec0re) [inline libr/util/num.c]: Wouldn't it be better to have an additional function (r_num_as_string_ex) to keep the API stable? r_num_as_string could simply call this funtion with default parameters (disabling escapes in this case). Or what's your current approach regarding API changes?

**Comment by radare** [comment]
hello? can you update the pr?

**Comment by bluec0re** [comment]
Any comment to my suggestions?

**Comment by radare** [comment]
I prefer a parameter and i dont think this function is called from many places

File: libr/util/num.c (hunk 2 above)
Comment (radare) [inline libr/util/num.c]: i would do a new parameter

**Comment by radare** [comment]
uhm i though i have already replied that. wtf

**Comment by bluec0re** [comment]
OK. I'll adjust it tomorrow. Have no access to a computer right now.

File: libr/util/num.c [hunk 3]
```diff
@@ -428,6 +449,8 @@ R_API char* r_num_as_string(RNum *___, ut64 n) {
 	}
 	if (ret)
 		return strdup (str);
```
Comment (radare) [inline libr/util/num.c]: Use braces even for single line blocks

File: libr/util/num.c [hunk 4]
```diff
@@ -408,17 +408,38 @@ R_API ut64 r_num_get_input_value(RNum *num, const char *input_value) {
...
 	int len = sizeof (ut64);
 	ut64 num = n;
 	str[stri=0] = 0;
 	while (len--) {
 		char ch = (num & 0xff);
-		if (ch>=33 && ch <127) {
+		if (ch>=32 && ch <127) {
 			str[stri++] = ch;
 			str[stri] = 0;
+		} else if(!printable_only && (off = escape_char (str + stri, ch)) != 0) {
```
Comment (radare) [inline libr/util/num.c]: Missing space before (

File: libr/util/num.c [hunk 5]
```diff
@@ -408,17 +408,38 @@ R_API ut64 r_num_get_input_value(RNum *num, const char *input_value) {
...
+R_API char* r_num_as_string(RNum *___, ut64 n, bool printable_only) {
+	char str[34]; // 8 byte * 4 chars in \x?? format
+	int stri, ret = 0, off = 0;
 	int len = sizeof (ut64);
 	ut64 num = n;
 	str[stri=0] = 0;
 	while (len--) {
 		char ch = (num & 0xff);
-		if (ch>=33 && ch <127) {
+		if (ch>=32 && ch <127) {
```
Comment (radare) [inline libr/util/num.c]: Missing spaces around >=

**Pull request merged**